### PR TITLE
okhttp: properly verify IPv6 address hosts

### PR DESCRIPTION
**Pull request by OnlyInAmerica**
Address mismatch between IPv6 address hosts derived from URIs and X509 subjectAltName extensions as discussed in #4278

### CODE REVIEW

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpTlsUpgrader.java [hunk 1]
```diff
@@ -70,9 +70,24 @@ public static SSLSocket upgrade(SSLSocketFactory sslSocketFactory,
...
+   *
+   * <p>IPv6 host addresses derived from URIs are enclosed in square brackets per RFC2732, but
+   * omit these brackets in X509 certificate subjectAltName extensions per RFC5280.
+   *
+   * @see <a href="https://www.ietf.org/rfc/rfc2732.txt">RFC2732</a>
+   * @see <a href="https://tools.ietf.org/html/rfc5280#section-4.2.1.6">RFC5280</a>
+   *
+   * @return {@param host} in a form consistent with X509 certificates
+   */
+  static String canonicalizeHost(String host) {
```
Comment (ericgribkoff) [inline okhttp/src/main/java/io/grpc/okhttp/OkHttpTlsUpgrader.java]: This can be private

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpTlsUpgrader.java [hunk 2]
```diff
@@ -70,9 +70,24 @@ public static SSLSocket upgrade(SSLSocketFactory sslSocketFactory,
...
+    return host.replaceAll("[\\[\\]]", "");
```
Comment (ericgribkoff) [inline okhttp/src/main/java/io/grpc/okhttp/OkHttpTlsUpgrader.java]: This may be safe as-is, but we should be more conservative here: if the first and last characters are `[` and `]`, strip only those two characters out. Otherwise, just return the string.

This would mean that we can add a test like `assertEquals("[someIp]", canonicalizeHost("[[someIp]]"));`. This address would, as in #4278, result in `OkHostnameVerifier` failing to match the host - but that's good, because having two brackets is malformed anyways, and probably indicates a bug somewhere else that we don't want to mask here.

File: okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java [hunk 3]
```diff
@@ -32,4 +34,44 @@
...
+    assertEquals(canonicalizeHost("1080:0:0:0:8:800:200C:417A"), "1080:0:0:0:8:800:200C:417A");
+    assertEquals(canonicalizeHost("[1080:0:0:0:8:800:200C:417A]"), "1080:0:0:0:8:800:200C:417A");
+    assertEquals(canonicalizeHost("1080::8:800:200C:417A"), "1080::8:800:200C:417A");
+    assertEquals(canonicalizeHost("[1080::8:800:200C:417A]"), "1080::8:800:200C:417A");
+    assertEquals(canonicalizeHost("FF01::101"), "FF01::101");
+    assertEquals(canonicalizeHost("[FF01::101]"), "FF01::101");
+    assertEquals(canonicalizeHost("0:0:0:0:0:0:13.1.68.3"), "0:0:0:0:0:0:13.1.68.3");
+    assertEquals(canonicalizeHost("[0:0:0:0:0:0:13.1.68.3]"), "0:0:0:0:0:0:13.1.68.3");
+    assertEquals(canonicalizeHost("0:0:0:0:0:FFFF:129.144.52.38"), "0:0:0:0:0:FFFF:129.144.52.38");
+    assertEquals(canonicalizeHost("[0:0:0:0:0:FFFF:129.144.52.38]"), "0:0:0:0:0:FFFF:129.144.52.38");
```
Comment (ericgribkoff) [inline okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java]: I think this fails checkstyle (100 char limit per line)

File: okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java [hunk 4]
```diff
@@ -32,4 +34,44 @@
             || OkHttpTlsUpgrader.TLS_PROTOCOLS.indexOf(Protocol.GRPC_EXP)
                 < OkHttpTlsUpgrader.TLS_PROTOCOLS.indexOf(Protocol.HTTP_2));
   }
+
+  @Test public void upgrade_canonicalizeHost() {
+    // IPv4
+    assertEquals(canonicalizeHost("127.0.0.1"), "127.0.0.1");
```
Comment (ericgribkoff) [inline okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java]: The argument order for `assertEquals` is reversed. These should be in the form: `assertEquals("ip", canonicalizeHost("ip"))`. (this results in better output when a test fails, as error message will label the first argument as the expected value and the second argument as the actual value)

File: okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java [hunk 5]
```diff
@@ -32,4 +34,44 @@
             || OkHttpTlsUpgrader.TLS_PROTOCOLS.indexOf(Protocol.GRPC_EXP)
                 < OkHttpTlsUpgrader.TLS_PROTOCOLS.indexOf(Protocol.HTTP_2));
   }
+
+  @Test public void upgrade_canonicalizeHost() {
```
Comment (ericgribkoff) [inline okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java]: Tests coverage look good to me (with the addition of a case to ensure we don't strip out additional brackets for malformed URIs). Out of curiosity, what methodology went into choosing the test hosts?

Comment (OnlyInAmerica) [reply okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java]: Will update to ensure we don't strip out additional brackets. 

I lifted these from okhttp's [HostnameVerifierTest](https://github.com/square/okhttp/blob/master/okhttp-tests/src/test/java/okhttp3/internal/tls/HostnameVerifierTest.java#L519), so I'll add some attribution.

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpTlsUpgrader.java (hunk 1 above)
Comment (OnlyInAmerica) [reply okhttp/src/main/java/io/grpc/okhttp/OkHttpTlsUpgrader.java]: Is package-private with `@VisibleForTesting` (ala `TLS_PROTOCOLS`) acceptable?

Comment (ericgribkoff) [reply okhttp/src/main/java/io/grpc/okhttp/OkHttpTlsUpgrader.java]: Oops, my mistake - yes, the visibility is fine, but the `@VisibleForTesting` annotation would be helpful.

File: okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java (hunk 5 above)
Comment (ericgribkoff) [reply okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java]: Let's simplify the tests somewhat here, which will also remove the need for attribution - we actually don't care about all of the edge cases that the linked test case covers, because we don't need to distinguish between IP addresses and hostnames. Probably just the following are enough:

```
assertEquals("::1", canonicalizeHost("::1"));
assertEquals("::1", canonicalizeHost("[::1]"));
assertEquals("127.0.0.1", canonicalizeHost("127.0.0.1"));
assertEquals("some.long.url.com", canonicalizeHost("some.long.url.com"));

// Extra square brackets in a malformed URI are retained
assertEquals("[::1]", canonicalizeHost("[[::1]]"));
```

Also, does it make sense to rename the test method to just `canonicalizeHost`, as this isn't testing any other part of the upgrade logic?

Comment (OnlyInAmerica) [reply okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java]: 👍 Is `canonicalizeHosts` ok to avoid clash with the static import of `canonicalizeHost`?

Comment (ericgribkoff) [reply okhttp/src/test/java/io/grpc/okhttp/OkHttpTlsUpgraderTest.java]: sounds good to me

**Comment by OnlyInAmerica** [comment]
fixed whitespace issue to appease travis. One more time from the top 🥁

**Comment by ericgribkoff** [comment]
Thanks!
